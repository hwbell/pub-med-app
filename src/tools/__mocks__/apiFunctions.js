const fetch = require('node-fetch')

// query and return based on date / citations
export async function getArticles(query, sortParam) {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  // let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;

  // let headers = {}
  // headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let list = [];
  for (let i = 0; i < 50; i++) {
    list.push({
      title: `Article ${i + 1}`,
      pmid: `PMID10000${i + 1}`
    })
  }

  let articles = await Promise.resolve({
    resultList: {
      result: list
    }
  })

  return articles;
}

// parses the result of a search query to a simple array of the article titles
export function parseSearchToTitlesArray(searchResults) {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })

  return titlesArray;
}