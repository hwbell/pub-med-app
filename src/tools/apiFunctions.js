const fetch = require('node-fetch')

// query and return based on date / citations
const getArticles = async (query, sortParam) => {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;
  
  let headers = {}
  headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let articles = await fetch(searchUrl, {
    method: 'POST',
    headers
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return articles;
}

// parses the result of a search query to a simple array of the article titles
const parseSearchToTitlesArray = (searchResults) => {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })
  return titlesArray;
}

// let results = getArticles('cancer').then((response) => {
//   console.log(response.resultList.result.slice(0,2))
// })

module.exports = {
  getArticles,
  parseSearchToTitlesArray,
}
