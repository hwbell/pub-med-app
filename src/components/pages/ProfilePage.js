import React from 'react';
import '../../App.scss';
import 'bootstrap/dist/css/bootstrap.css';
import { CSSTransitionGroup } from 'react-transition-group' // ES6

// components
import Header from '../Header';
import ProfileForm from '../ProfileForm';
import { Button, Form, FormGroup, Label, Input, Fade } from 'reactstrap';

// tools
import { signInUser, patchUser } from '../../tools/serverFunctions';

// ******************************************************************************
class ProfilePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      checked: false,
      editText: false,
      showProfileForm: false
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleCheck = this.handleCheck.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleSubmitUser = this.handleSubmitUser.bind(this);
    this.attemptSignIn = this.attemptSignIn.bind(this);
    this.renderProfile = this.renderProfile.bind(this);
    this.toggleEditText = this.toggleEditText.bind(this);
    this.toggleProfileForm = this.toggleProfileForm.bind(this);
  }

  componentDidMount() {
    // console.log(localStorage)
  }

  renderSignIn() {
    return (
      <div className="outline signin" style={{ padding: '40px' }}>
        <p className="paragraph">
          {`Sign in or create an account to start sharing resources with other researchers!`}
        </p>

        <p className="paragraph">
          {`If you don't already have an account, you can sign up by checking the new user box below
          and confirming your password.`}
        </p>

        {this.renderInputs()}
      </div>
    )
  }

  renderInputs() {
    return (
      <Form style={styles.form} onSubmit={this.handleSubmit}>

        <Input style={styles.input} type="email" name="email" placeholder="email"
          onChange={(e) => this.handleChange(e)}
        />
        <Input style={styles.input} type="password" name="password" placeholder="password"
          onChange={(e) => this.handleChange(e)}
        />

        <CSSTransitionGroup
          transitionName="fade"
          transitionEnterTimeout={500}
          transitionLeaveTimeout={300}>

          {this.state.checked &&
            <Input key="confirm-password" style={styles.input} type="password" name="confirm-password" placeholder="confirm password"
              onChange={(e) => this.handleChange(e, true)}
            />}

          {this.state.checked &&
            <Input style={styles.input} type="text" name="name" placeholder="username (optional)"
              onChange={(e) => this.handleChange(e)}
            />}


        </CSSTransitionGroup>

        <FormGroup style={{ marginTop: '8px' }} check>
          <Label check>
            <Input type="checkbox" checked={this.state.checked}
              onChange={this.handleCheck} />
            <p>new user?</p>
          </Label>
        </FormGroup>

        <Button color="primary" type="submit">sign up!</Button>
      </Form>
    )
  }

  renderProfile() {

    let user = this.props.user;
    console.log(user)

    if (!user) {
      return;
    }

    return (
      <div className="outline profile" style={styles.profileInfo}>

        <Button className="warn article-button" style={styles.button}
          onClick={this.props.registerSignOut}>logout</Button>

        <div style={styles.nameAndIcon}>
          <p className="profile-title">
            <strong>{`${this.props.user.name || this.props.user.email}`}</strong>
          </p>

          <i className="profile-title fas fa-user-edit" style={styles.icon}
            onMouseOver={this.toggleEditText}
            onMouseLeave={this.toggleEditText}
            onClick={this.toggleProfileForm}></i>

          <Fade in={this.state.editText} className="profile-title">
            <p style={{ color: 'white' }}>
              add to your profile!
            </p>
          </Fade>
        </div>

        <p className="profile-title">
          {user.email}
        </p>

        <p className="profile-title">
          {user.collections &&
            `${user.collections.length} collections`}
        </p>
      </div>
    )
  }

  handleChange(e, confirming) {

    let { name, value } = e.target;

    let obj = this.state.user || {};
    obj[`${name}`] = value;

    this.setState({ user: obj }, () => {
      // console.log(this.state.user)
    })
  }

  handleCheck() {
    this.setState({
      checked: !this.state.checked
    })
  }

  // this one is for the Login Form
  handleSubmit(e) {
    e.preventDefault()
    if (!this.state.user) {
      return;
    }

    // check for all fields, with username being optional
    let user = this.state.user;
    if (!user.email || !user.password) {
      return console.log('Email and password are required!');
    }

    // check for confirmed password if new user
    if (this.state.checked) {
      if (user.password !== user['confirm-password']) {
        return console.log('passwords dont match, you new user you!')
      }
    }

    console.log(`input looks ok for user: ${user.name}`)

    return this.attemptSignIn(user);
  }

  attemptSignIn(user) {

    // sign the user in via post to the server. get the token and store it for future requests
    // pass this.state.checked as a boolean for whether it is a new user
    return signInUser(user, this.state.checked)
      .then((response) => {
        console.log(response)

        let { user, token } = response;

        // save info to local storage
        localStorage.setItem(`user`, JSON.stringify(user));
        localStorage.setItem(`token`, JSON.stringify(token));

        // register back in App
        return this.props.registerSignIn(user);

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  toggleEditText() {
    this.setState({
      editText: !this.state.editText
    })
  }

  toggleProfileForm() {
    this.setState({
      showProfileForm: !this.state.showProfileForm
    })
  }

  // this one is for the Profile Form
  handleSubmitUser(profileInfo) {
    if (!profileInfo || Object.keys(profileInfo).length === 0) {
      return;
    }

    // filter out any empty entries
    let patch = {};
    Object.keys(profileInfo).forEach((key) => {
      if (info[key]) {
        patch[key] = info[key]
      }
    })

    let token = JSON.parse(localStorage.getItem('token'));
    let headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }

    // Send the patch to save it to the server. Using the currently stored token
    // will assign the user as the owner of the collection in mongodb on the backend
    return patchUser(patch, headers)
      .then((response) => {
        console.log(response)

        // once we have successfully patched, we will: 

        // 1. Refresh the user - this will send the new collection list sent back
        // from the server to the root App component to update all concerned components
        this.props.refreshUser(response);

        // clear the info - we only need it in state before its sent to the server
        this.setState({
          profileInfo: {}
        })

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }


  render() {

    // console.log(this.props)
    return (
      <div className="page">

        {/* the edit modal, for user that is logged in */}
        <ProfileForm
          isVisible={this.state.showProfileForm}
          user={this.props.user}
          toggle={this.toggleProfileForm}
          handleSubmitUser={this.handleSubmitUser}
        />

        <div className="glass">

          <Header
            class="heading"
            title={"Your Profile"}
            subtitle={""}
          />

          {!this.props.user ?
            this.renderSignIn() :
            this.renderProfile()
          }

        </div>

      </div>
    );
  }
}

const styles = {
  form: {
    width: '80%',
    maxWidth: '300px',
    margin: 0,
    padding: 0,
  },
  input: {
    margin: 0,
    marginTop: '8px'
  },
  button: {
    alignSelf: 'flex-end'
  },
  profileInfo: {
    padding: '20px',
    alignItems: 'flex-start'
  },
  nameAndIcon: {
    display: 'flex',
    flexDirection: 'row',
  }
}

export default ProfilePage;
