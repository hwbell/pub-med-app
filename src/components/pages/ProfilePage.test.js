import React from 'react';
import ProfilePage from './ProfilePage';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { getArticles } from '../../tools/apiFunctions.js';

// sample props
const user = {
  name: 'Harry',
  email: 'harry@example.com'
};

const newThread = {
  name: 'Figure 5 in PMID: 83470572093',
  article: '83470572093'
}

describe('ProfilePage', () => {
  
  let articles;
  let threads;
  let collections;
  let someProps;

  beforeAll( async () => {
    articles = [
      {
        id: '013091283'
      },
      {
        id: '446264632'
      },
      {
        id: '651753334'
      }
    ]

    threads = [
      {
        name: 'Sourcing of PMID: 013091283',
        article: '013091283'
      },
      {
        name: 'Data in figure 1 of PMID: 013091283',
        article: '013091283'
      },
      {
        name: 'Figure 3 in PMID: 013091283',
        article: '013091283'
      }
    ]

    collections = [
      {
        name: 'first collection',
        articles: articles
      },
      {
        name: 'second collection',
        articles: articles
      }
    ]

    someProps = {
      user: {
        name: 'Mark',
        email: 'mark@mail.com',
        age: 33,
        threads,
        collections
      }
    }
  }) 

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ProfilePage {...someProps}/>);
  })

  // tests
  it('renders without crashing', async () => {
    wrapper.update();

  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<ProfilePage />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {
    let wrapper = mount(<ProfilePage />);

    ['.page', '.page-content', '.glass', '.heading', '.outline', 'Button'].forEach((selector) => {
      expect(wrapper.find(selector).length).toEqual(1);
    });

    expect(wrapper.find('AlertModal').length).toBe(1);

  });

  it('shows the signin form if user is not signed in', () => {
    
    wrapper.update();
    expect(wrapper.find('.profile').length).toEqual(1);
    expect(wrapper.find('.signin').length).toEqual(0);
    expect(wrapper.find('.profile-title').length).toEqual(1);
    expect(wrapper.find('.thread-text').length).toEqual(3);

    expect(wrapper.find('Button').length).toEqual(2);    
    expect(wrapper.find('.fa-user-edit').length).toEqual(1);    
    expect(wrapper.find('.fa-sign-out-alt').length).toEqual(1);    

    expect(wrapper.find('Fade').length).toBe(1);

    // change the signedIn prop and wipe the user
    wrapper.setProps({ user: null });
    wrapper.update();
    // check for the change
    expect(wrapper.find('.profile').length).toEqual(0);
    expect(wrapper.find('.signin').length).toEqual(1);
    expect(wrapper.find('.profile-title').length).toEqual(0);
    expect(wrapper.find('Button').at(0).render().text()).toEqual('sign up!');
    expect(wrapper.find('Fade').length).toBe(0);
  })

  it('should show the users profile once signed in', async () => {
    let wrapper = mount(<ProfilePage />);

    // at first, no profile
    expect(wrapper.find('.profile').length).toEqual(0);

    // change the state and check for its appearance. need to supply
    // a user as well

    wrapper.setProps({ signedIn: true });
    wrapper.setProps({ user });

    wrapper.update();
    expect(wrapper.find('.profile').length).toEqual(1);
  })

  it('handleCheck() should toggle the checkbox', () => {
    let wrapper = mount(<ProfilePage />);
    expect(wrapper.state().checked).toBe(false);
    expect(wrapper.find('Input').length).toEqual(3);

    // check it
    wrapper.instance().handleCheck();
    wrapper.update();
    expect(wrapper.state().checked).toBe(true);
    expect(wrapper.find('Input').length).toEqual(5);

    // uncheck it
    wrapper.instance().handleCheck();
    wrapper.update();
    expect(wrapper.state().checked).toBe(false);
    // wait for the css transition out
    setTimeout(() => {
      expect(wrapper.find('Input').length).toEqual(3);
    }, 500)

  })

  it('toggleAlertModal() should toggle the AlertModal', async () => {
    let wrapper = await mount(<ProfilePage />);

    expect(wrapper.state().modalProps.isVisible).toBe(false);
    expect(wrapper.find('AlertModal').props().isVisible).toBe(false);

    wrapper.instance().toggleAlertModal();
    wrapper.update();

    expect(wrapper.state().modalProps.isVisible).toBe(true);   
    expect(wrapper.find('AlertModal').props().isVisible).toBe(true);

    wrapper.instance().toggleAlertModal();
    wrapper.update();

    expect(wrapper.state().modalProps.isVisible).toBe(false);   
    expect(wrapper.find('AlertModal').props().isVisible).toBe(false);

    
  })

  it('togglePopup should assign this.state.popupText and toggle the boolean for the Fade', () => {
    expect(wrapper.state().showPopup).toBe(false);
    expect(wrapper.find('Fade').props().in).toBe(false);
    expect(wrapper.find('Fade').render().text()).toBe('');

    wrapper.instance().togglePopup('fade text');
    wrapper.update();

    expect(wrapper.find('Fade').props().in).toBe(true);
    expect(wrapper.find('Fade').render().text()).toBe('fade text');
    expect(wrapper.state().showPopup).toBe(true);
    
  })

  it('should show the Fade text when the edit / logout buttons are hovered', async () => {
    // mount to check inner props of the Fade
    let wrapper = await mount(<ProfilePage {...someProps}/>);
    
    // start not showing the fade
    expect(wrapper.state().showPopup).toBe(false);    
    expect(wrapper.find('Fade').props().in).toBe(false);

    // toggles when the user hovers on the edit button
    wrapper.find('.fa-user-edit').simulate('mouseover');
    wrapper.update();
    
    expect(wrapper.state().showPopup).toBe(true);
    expect(wrapper.find('Fade').props().in).toBe(true);
    expect(wrapper.find('Fade').render().text()).toBe('edit profile');

     // toggles when the user hovers on the logout button
     wrapper.find('.fa-sign-out-alt').simulate('mouseover');
     wrapper.update();
     
     expect(wrapper.state().showPopup).toBe(true);
     expect(wrapper.find('Fade').props().in).toBe(true);
     expect(wrapper.find('Fade').render().text()).toBe('log out');

  })

  it('toggles the ProfileForm modal when the edit icon is clicked', () => {

    expect(wrapper.state().showProfileForm).toBe(false);
    
    wrapper.instance().toggleProfileForm();
    wrapper.update();
    expect(wrapper.state().showProfileForm).toBe(true);

    wrapper.instance().toggleProfileForm();    
    wrapper.update();
    expect(wrapper.state().showProfileForm).toBe(false);

  })

  it('should show atom icons representing the number of user collections', () => {

    // we gave the user 2 collections
    expect(wrapper.find('.fa-atom').length).toBe(2);

    // add another collection
    let user = someProps.user;
    user.collections.push(articles[0]);
    wrapper.setProps({user});
    wrapper.update();
   
    expect(wrapper.find('.fa-atom').length).toBe(3);

  })

  it('should show dna icons representing the number of user threads', () => {

    // we gave the user 3 threads
    expect(wrapper.find('.fa-dna').length).toBe(3);

    // add another thread
    let user = someProps.user;
    user.threads.push(newThread);
    wrapper.setProps({user});
    wrapper.update();
   
    expect(wrapper.find('.fa-dna').length).toBe(4);
  })
  

})
