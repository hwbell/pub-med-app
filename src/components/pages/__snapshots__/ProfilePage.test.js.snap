// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage renders correctly 1`] = `
<div
  className="page"
>
  <div
    className="glass page-content"
  >
    <div
      className="heading"
    >
      <h2
        className="title"
      >
        Your Profile
      </h2>
      <h4
        className="subtitle"
      >
        say something about you
      </h4>
    </div>
    <div
      className="outline signin"
      style={
        Object {
          "padding": "40px",
        }
      }
    >
      <p
        className="paragraph"
      >
        Sign in or create an account to start sharing resources with other researchers!
      </p>
      <p
        className="paragraph"
      >
        If you don't already have an account, you can sign up by checking the new user box below
          and confirming your password.
      </p>
      <form
        className=""
        onSubmit={[Function]}
        style={
          Object {
            "margin": 0,
            "maxWidth": "300px",
            "padding": 0,
            "width": "80%",
          }
        }
      >
        <input
          className="form-control"
          name="email"
          onChange={[Function]}
          placeholder="email"
          style={
            Object {
              "margin": 0,
              "marginTop": "8px",
            }
          }
          type="email"
        />
        <input
          className="form-control"
          name="password"
          onChange={[Function]}
          placeholder="password"
          style={
            Object {
              "margin": 0,
              "marginTop": "8px",
            }
          }
          type="password"
        />
        <span />
        <div
          className="form-check"
          style={
            Object {
              "marginTop": "8px",
            }
          }
        >
          <label
            className="form-check-label"
          >
            <input
              checked={false}
              className="form-check-input"
              onChange={[Function]}
              type="checkbox"
            />
            <p>
              new user?
            </p>
          </label>
        </div>
        <button
          aria-label={null}
          className="btn btn-primary"
          onClick={[Function]}
          type="submit"
        >
          sign up!
        </button>
      </form>
    </div>
  </div>
</div>
`;
